redis/src/connection.rs

```diff
@@ -918,71 +918,195 @@ pub(crate) fn create_rustls_config(
     }
 }
 
-fn connect_auth(
-    con: &mut Connection,
+fn authenticate_cmd(
     connection_info: &RedisConnectionInfo,
+    check_username: bool,
     password: &str,
-) -> RedisResult<()> {
+) -> Cmd {
     let mut command = cmd("AUTH");
-    if let Some(username) = &connection_info.username {
-        command.arg(username);
-    }
-    let err = match command.arg(password).query::<Value>(con) {
-        Ok(Value::Okay) => return Ok(()),
-        Ok(_) => {
-            fail!((
-                ErrorKind::ResponseError,
-                "Redis server refused to authenticate, returns Ok() != Value::Okay"
-            ));
+    if check_username {
+        if let Some(username) = &connection_info.username {
+            command.arg(username);
         }
-        Err(e) => e,
-    };
-    let err_msg = err.detail().ok_or((
-        ErrorKind::AuthenticationFailed,
-        "Password authentication failed",
-    ))?;
-    if !err_msg.contains("wrong number of arguments for 'auth' command") {
-        fail!((
-            ErrorKind::AuthenticationFailed,
-            "Password authentication failed",
-        ));
-    }
-
-    // fallback to AUTH version <= 5
-    let mut command = cmd("AUTH");
-    match command.arg(password).query::<Value>(con) {
-        Ok(Value::Okay) => Ok(()),
-        _ => fail!((
-            ErrorKind::AuthenticationFailed,
-            "Password authentication failed",
-        )),
     }
+    command.arg(password);
+    command
 }
 
 pub fn connect(
     connection_info: &ConnectionInfo,
     timeout: Option<Duration>,
 ) -> RedisResult<Connection> {
-    let con = ActualConnection::new(&connection_info.addr, timeout)?;
+    let con: ActualConnection = ActualConnection::new(&connection_info.addr, timeout)?;
     setup_connection(con, &connection_info.redis)
 }
 
-#[cfg(not(feature = "disable-client-setinfo"))]
-pub(crate) fn client_set_info_pipeline() -> Pipeline {
-    let mut pipeline = crate::pipe();
+pub(crate) struct ConnectionSetupComponents {
```

Comment:
I get the point of the struct, but I wonder if we can somehow live w/out it? Maybe at least shorten the field names to make it less noticeable :joy:

Reply:
I like the explicitly long names :(
Maybe 
```rs
enum CmdIndex {
  Resp2Auth,
  Resp3Auth,
  SelectDb
}

type  ConnectionSetupComponents = HashMap<CmdIndex, usize>;
```
?

Reply:
I would just make the fields shorter.
```
pub(crate) struct ConnectionSetupComponents {
    resp3_auth_cmd_idx: Option<usize>,
    resp2_auth_cmd_idx: Option<usize>,
    select_cmd_idx: Option<usize>,
}
```